File: CSharp/ISP/After.cs
// Smaller interfaces
public interface IPrinter
{
    void Print(string document);
}

public interface IFaxMachine
{
    void Fax(string document);
}

// Printer class now implements only the printer interface
public class MyPrinter : IPrinter
{
    public void Print(string document)
    {
        Console.WriteLine($"Printing: {document}");
    }
}

// Fax machine that implements the faxing interface
public class MyFaxMachine : IFaxMachine
{
    public void Fax(string document)
    {
        Console.WriteLine($"Faxing: {document}");
    }
}

// Client code
public class MyOffice
{
    public void ProcessDocument(IPrinter printer, string document)
    {
        printer.Print(document);
    }

    public void SendFax(IFaxMachine faxMachine, string document)
    {
        faxMachine.Fax(document);
    }
}

File: CPlusPlus/ISP/After.cpp
// MyMachine.cpp
#include <iostream>
#include <string>

// Interface for printing
class IPrinter {
public:
    virtual void Print(const std::string& document) = 0;
    virtual ~IPrinter() = default;
};

// Interface for faxing
class IFaxMachine {
public:
    virtual void Fax(const std::string& document) = 0;
    virtual ~IFaxMachine() = default;
};

// Concrete printer
class MyPrinter : public IPrinter {
public:
    void Print(const std::string& document) override {
        std::cout << "Printing: " << document << std::endl;
    }
};

// Concrete fax machine
class MyFaxMachine : public IFaxMachine {
public:
    void Fax(const std::string& document) override {
        std::cout << "Faxing: " << document << std::endl;
    }
};

// Client code
class MyOffice {
public:
    void ProcessDocument(IPrinter& printer, const std::string& document) {
        printer.Print(document);
    }

    void SendFax(IFaxMachine& faxMachine, const std::string& document) {
        faxMachine.Fax(document);
    }
};
